Your job is to build a funnel for CoolTShirts.com, which is a description of how many people continue to the next step of a multi-step process.

In this case, our funnel is going to describe the following process:

<ol>
  <li>A user visits CoolTShirts.com</li>
  <li>A user adds a t-shirt to their cart</li>
  <li>A user clicks "checkout"</li>
  <li>A user actually purchases a t-shirt</li>
</ol>

In [40]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

#import the relevant datasets
visits = pd.read_csv('visits.csv',parse_dates=[1])
cart = pd.read_csv('cart.csv', parse_dates=[1])
checkout = pd.read_csv('checkout.csv', parse_dates=[1])
purchase = pd.read_csv('purchase.csv',parse_dates=[1])

Next, we shall explore every DataFrame before further action.

In [2]:
print(visits.head())

                                user_id          visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00


In [3]:
print(cart.head())

                                user_id           cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3 2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00


In [4]:
print(checkout.head())

                                user_id       checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00
3  89fe330a-8966-4756-8f7c-3bdbcd47279a 2017-04-20 16:15:00
4  3ccdaf69-2d30-40de-b083-51372881aedd 2017-01-08 20:52:00


In [5]:
print(purchase.head())

                                user_id       purchase_time
0  4b44ace4-2721-47a0-b24b-15fbfa2abf85 2017-05-11 04:25:00
1  02e684ae-a448-408f-a9ff-dcb4a5c99aac 2017-09-05 08:45:00
2  4b4bc391-749e-4b90-ab8f-4f6e3c84d6dc 2017-11-20 20:49:00
3  a5dbb25f-3c36-4103-9030-9f7c6241cd8d 2017-01-22 15:18:00
4  46a3186d-7f5a-4ab9-87af-84d05bfd4867 2017-06-11 11:32:00


#### Left Merge DataFrame Visits and Cart

In [7]:
visit_cart = pd.merge(visits, cart, how = 'left')

print(visit_cart.head())

                                user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
4  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   

            cart_time  
0                 NaT  
1 2017-01-26 14:44:00  
2 2017-08-20 08:31:00  
3 2017-08-20 08:49:00  
4                 NaT  


In [9]:
#Length of the left merge DataFrame - number of visitors
print(len(visit_cart))

2052


In [31]:
#Records that show Null under column "cart_time"
null_cart = visit_cart[visit_cart['cart_time'].isnull()]
                       
visit_cart[visit_cart['cart_time'].isnull()].head()

user_id          visit_time cart_time
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00       NaT
4  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00       NaT
5  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00       NaT
6  7bc62113-d51d-4e6b-85e0-1cf29e3de74a 2017-06-03 20:05:00       NaT
7  6f22cd44-bc63-4449-a9af-2651859da2f9 2017-03-18 23:09:00       NaT

Caculating the percent of users who visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart.

In [46]:
percent_no_cart = len(null_cart) / len(visit_cart) * 100

print("The percent of users visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart is {:.2f}%.".format(percent_no_cart))

The percent of users visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart is 80.51%.


Next, we will investigate the check out percentage for visitors who placed a t-shirt in their cart.

In [41]:
cart_checkout = pd.merge(cart, checkout,how='left')
print(cart_checkout.head())


                                user_id           cart_time  \
0  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
1  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
2  2be90e7c-9cca-44e0-bcc5-124b945ff168 2017-11-07 20:45:00   
3  4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00   
4  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663 2017-03-04 10:38:00   

        checkout_time  
0 2017-11-07 21:14:00  
1 2017-11-07 20:50:00  
2 2017-11-07 21:11:00  
3                 NaT  
4 2017-03-04 11:04:00  


In [42]:
#number of visitors who placed a cart.
cart_checkout_row = len(cart_checkout)

#Create a DataFrame for visitors who put a t-shirt into Cart but they did not proceed to checkout.
cart_check_null = cart_checkout[cart_checkout['checkout_time'].isnull()]

cart_check_null.head()

user_id           cart_time checkout_time
3   4397f73f-1da3-4ab3-91af-762792e25973 2017-05-27 01:35:00           NaT
6   a68a16e2-94f0-4ce8-8ce3-784af0bbb974 2017-07-26 15:48:00           NaT
14  fd80ce93-ae6e-4c0b-9ea4-561f84152026 2017-06-07 01:18:00           NaT
26  48a23075-694b-417d-8449-9df921ad95aa 2017-07-09 15:28:00           NaT
29  5d7d121a-817c-4b84-b4d6-5388092b9aec 2017-06-10 14:54:00           NaT

Calculating the percentage of users put items in their cart, but did not proceed to checkout.

In [45]:
percent_no_checkout = len(cart_check_null) / cart_checkout_row  * 100

print("The percentage of users put items in their cart, but did not proceed to checkout is {:.2f}%.".format(percent_no_checkout))

The percentage of users put items in their cart, but did not proceed to checkout is 20.93%.


#### Merge the checkout and purchase DataFrame

In [72]:
checkout_purchase = checkout.merge(purchase, how='left')
checkout_purchase.head()

user_id       checkout_time  \
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00   
1  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00   
2  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d 2017-06-25 09:29:00   
3  4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00   
4  3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00   

        purchase_time  
0 2017-06-25 09:49:00  
1 2017-06-25 09:56:00  
2 2017-06-25 09:57:00  
3                 NaT  
4                 NaT

Calculate the percentage of users proceeded to checkout, but did not purchase a t-shirt.

In [73]:
#Create a DataFrame which purchase_time is NULL
purchase_null = checkout_purchase[checkout_purchase['purchase_time'].isnull()]

purchase_null.head()

user_id       checkout_time purchase_time
3   4ac186f0-9954-4fea-8a27-c081e428e34e 2017-04-07 20:11:00           NaT
4   3c9c78a7-124a-4b77-8d2e-e1926e011e7d 2017-07-13 11:38:00           NaT
8   a25aad6a-55af-4ddc-92d1-e8c58f855329 2017-01-05 20:07:00           NaT
9   3327f8fb-c5a0-421e-a0d1-fb619b79ab45 2017-08-17 04:20:00           NaT
27  bba9bf5b-5817-4371-b328-c3764db7f1ae 2017-02-22 17:25:00           NaT

In [74]:
percent_no_purchase = len(purchase_null) / len(checkout_purchase) * 100

print("The percentage of users proceeded to checkout, but did not purchase a t-shirt is {:.2f}%.".format(percent_no_purchase))

The percentage of users proceeded to checkout, but did not purchase a t-shirt is 16.89%.


In [75]:
#A list 
funnel_percent = {'Cart':percent_no_cart,'Checkout':percent_no_checkout,'Purchase':percent_no_purchase}

funnel_df = pd.DataFrame(funnel_percent,index=[0])

#Percentage for all the funnel steps
print(funnel_df)

        Cart   Checkout   Purchase
0  80.506823  20.930233  16.889632


The Purchase step is the weakest. Only 16.9% of those who are at the checkout page eventually purchased the t-shirt.

We can focus on improving this conversion rate.

In [81]:
#Merge all four DataFrames
all_data = visits.merge(cart, how='left').merge(checkout, how='left').merge(purchase, how='left')

#Time from initial visit to the final purchase
all_data['time_to_purchase'] = all_data.purchase_time - all_data.visit_time

all_data.time_to_purchase.mean()

Timedelta('0 days 00:44:02.672413')

From the result, the average time taken from visiting the website to a purchase is 44 minutes.
We can improve on the timing, though this might not improve the conversion rate from checkout to purchase.